src/Processors/Formats/Impl/NpyRowInputFormat.cpp

```diff
@@ -0,0 +1,481 @@
+    }
+    throw Exception(ErrorCodes::UNKNOWN_TYPE, "Numpy type {} is not supported", magic_enum::enum_name(numpy_type->getTypeIndex()));
+}
+
+DataTypePtr createNestedArrayType(const DataTypePtr & nested_type, size_t depth)
+{
+    DataTypePtr result_type = nested_type;
+    assert(depth > 0);
+    if (depth > 1)
+    {
+        for (size_t i = 0; i < depth - 1; ++i)
+            result_type = std::make_shared<DataTypeArray>(std::move(result_type));
+    }
+    return result_type;
+}
+
+size_t parseTypeSize(const std::string type)
+{
+    try
+    {
+        size_t size = std::stoi(type);
+        return size;
+    }
+    catch (...)
+    {
+        throw Exception(ErrorCodes::INCORRECT_DATA, "Invalid data type");
+    }
+}
+
+std::shared_ptr<NumpyDataType> parseType(String type)
+{
+    /// Parse endianess
+    NumpyDataType::Endianness endianess;
+    if (type[0] == '<')
+        endianess = NumpyDataType::Endianness::LITTLE;
+    else if (type[1] == '>')
+        endianess = NumpyDataType::Endianness::BIG;
+    else if (type[0] == '|')
+        endianess = NumpyDataType::Endianness::NONE;
+    else
+      throw Exception(ErrorCodes::BAD_ARGUMENTS, "Wrong header data");
+
+    /// Parse type
+    if (type[1] == 'i')
+        return std::make_shared<NumpyDataTypeInt>(endianess, parseTypeSize(type.substr(2)), true);
+    else if (type[1] == 'b')
+        return std::make_shared<NumpyDataTypeInt>(endianess, parseTypeSize(type.substr(2)), false);
+    else if (type[1] == 'u')
+        return std::make_shared<NumpyDataTypeInt>(endianess, parseTypeSize(type.substr(2)), false);
+    else if (type[1] == 'f')
+        return std::make_shared<NumpyDataTypeFloat>(endianess, parseTypeSize(type.substr(2)));
+    else if (type[1] == 'S')
+        return std::make_shared<NumpyDataTypeString>(endianess, parseTypeSize(type.substr(2)));
+    else if (type[1] == 'U')
+        return std::make_shared<NumpyDataTypeUnicode>(endianess, parseTypeSize(type.substr(2)));
+    else if (type[1] == 'c')
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "ClickHouse doesn't support complex numeric type");
+    else if (type[1] == 'O')
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "ClickHouse doesn't support object types");
+    else
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Error while parsing data type");
+}
+
+std::vector<int> parseShape(String shape_string)
+{
+    shape_string.erase(std::remove(shape_string.begin(), shape_string.end(), '('), shape_string.end());
+    shape_string.erase(std::remove(shape_string.begin(), shape_string.end(), ')'), shape_string.end());
+    std::vector<std::string> result_str;
+    boost::split(result_str, shape_string, boost::is_any_of(","));
+
+    std::vector<int> shape;
+    if (result_str[result_str.size()-1].empty())
+        result_str.pop_back();
+    shape.reserve(result_str.size());
+    for (const String& item : result_str)
+        shape.push_back(std::stoi(item));
```

Comment:
`stoi` can throw an exception, and we want to make exception messaging better. Let's use our parsing function:
```suggestion
    for (const String & item : result_str)
    {
        shape.push_back();
        ReadBufferFromString buf(item);
        if (!tryReadIntText(shape.back(), buf))
            throw Exception(ErrorCodes::INCORRECT_DATA, "Invalid shape format: {}", shape_string);
    }
```